### ISSUE TITLE
Compiling from source mosquitto v1.5 on FreeBSD bug

### ISSUE THREAD

**Issue by denigmus** [eclipse/mosquitto#863]
git clone --branch v1.5 https://github.com/eclipse/mosquitto
cd mosquitto
gmake

```
cc sub_client.o sub_client_output.o client_shared.o -o mosquitto_sub  -L../lib ../lib/libmosquitto.so.1
gmake[1]: Leaving directory '/root/MOSQUITTO-UPDATE/mosquitto/client'
gmake[1]: Entering directory '/root/MOSQUITTO-UPDATE/mosquitto/src'
cc -Wall -ggdb -O2  -I. -I.. -I../lib  -DVERSION="\"1.5\"" -DWITH_BROKER -DWITH_TLS -DWITH_TLS_PSK -DWITH_BRIDGE -DWITH_PERSISTENCE -DWITH_MEMORY_TRACKING -DWITH_SYS_TREE -DWITH_EC -c mosquitto.c -o mosquitto.o
mosquitto.c:104:7: warning: implicit declaration of function 'initgroups' is invalid in C99 [-Wimplicit-function-declaration]
                        if(initgroups(config->user, pwd->pw_gid) == -1){
                           ^
mosquitto.c:231:2: warning: implicit declaration of function 'gettimeofday' is invalid in C99 [-Wimplicit-function-declaration]
        gettimeofday(&tv, NULL);
        ^
2 warnings generated.
cc -Wall -ggdb -O2  -I. -I.. -I../lib  -DVERSION="\"1.5\"" -DWITH_BROKER -DWITH_TLS -DWITH_TLS_PSK -DWITH_BRIDGE -DWITH_PERSISTENCE -DWITH_MEMORY_TRACKING -DWITH_SYS_TREE -DWITH_EC -c bridge.c -o bridge.o
cc -Wall -ggdb -O2  -I. -I.. -I../lib  -DVERSION="\"1.5\"" -DWITH_BROKER -DWITH_TLS -DWITH_TLS_PSK -DWITH_BRIDGE -DWITH_PERSISTENCE -DWITH_MEMORY_TRACKING -DWITH_SYS_TREE -DWITH_EC -c conf.c -o conf.o
conf.c:117:24: error: use of undeclared identifier 'PF_UNSPEC'
        gai_hints.ai_family = PF_UNSPEC;
                              ^
1 error generated.
gmake[1]: *** [Makefile:68: conf.o] Error 1
gmake[1]: Leaving directory '/root/MOSQUITTO-UPDATE/mosquitto/src'
gmake: *** [Makefile:26: mosquitto] Error 2

```
FreeBSD 11.1-RELEASE-p8 

gmake -v
GNU Make 4.2.1
Built for amd64-portbld-freebsd11.1

cc -v
FreeBSD clang version 4.0.0 (tags/RELEASE_400/final 297347) (based on LLVM 4.0.0)
Target: x86_64-unknown-freebsd11.1
Thread model: posix
InstalledDir: /usr/bin

**Comment by ptjm** [comment]
This is caused by defining _POSIX_C_SOURCE 200809 (in config.h). You could fix it by wrapping an #ifndef __FreeBSD__/#endif around that definition.

**Comment by karlp** [comment]
@ptjm say that again?  is some bsd compiler saying it's posix 200809 compatible, but it's not?  that's a bsd bug, not somethign that should be fixed here.

**Comment by denigmus** [comment]
@ptjm 
```
#else
#  define _DEFAULT_SOURCE 1
#   ifndef FreeBSD
#     define _POSIX_C_SOURCE 200809L
#   endif
#endif
```
Correct?
Doesn't help.

**Comment by ptjm** [comment]
@denigmus - sorry the markdown parser turned my text into formatting. There should be two underscores on either side of FreeBSD.
```
#else
#  define _DEFAULT_SOURCE 1
#   ifndef __FreeBSD__
#     define _POSIX_C_SOURCE 200809L
#   endif
#endif
```

@karlp - if you define _POSIX_C_SOURCE, you're supposed to write code which conforms to that version of the SUS. PF_UNSPEC is a non-standard BSDism and a conforming system will not define it if _POSIX_C_SOURCE is defined. That means using AF_UNSPEC, since that's what the standard defines, and while it's annoying that some systems don't define manifest constants which aren't in the standard while others do, what the standard says is that this is what is supposed to happen. Anyway, if you want to argue that mosquitto shouldn't compile on systems which don't define PF_UNSPEC in these circumstances, why not start with darwin, where _POSIX_C_SOURCE is deliberately not defined for exactly this reason in the #if part of the block cited above?

**Comment by karlp** [comment]
@ptjm if mosquitto is misusing posix source guards, then mosquitto should be fixed.  But fixed to use posix source properly.  Just saying that it's posix source for one OS, and non-posix source for another isn't fixing anything, it's just making a tangled mess.  From what you're saying, clearly, mosquitto should be using AF_UNSPEC, rather than hacking on these include guards.

**Comment by denigmus** [comment]
Compiled

**Comment by toast-uz** [comment]
Thanks. Shall I close this?

**Comment by denigmus** [comment]
Yep if no one wants to add this fix to the code ;)

**Comment by karlp** [comment]
@toast-uz NO! stop this!  what on earth makes you think this is resolved?!

**Comment by toast-uz** [comment]
Did I misunderstand @denigmus comment?

**Comment by toast-uz** [comment]
Sorry, I understood @karlp insisted mosquitto should not be fixed. and rejected @denigmus and @ptjm 's submitting fix, and @denigmus agreed. Misunderstood?

**Comment by denigmus** [comment]
I'm not a person to decide what to do with it. I don't even know C. I just said that @ptjm's fix has helped me to compile on FreeBSD.

**Comment by ptjm** [comment]
If I ever get around to dealing with the eclipse IP, I'll submit a pull request to get at the root issue -- replacing PF_* with AF_* to match POSIX.

**Comment by ralight** [comment]
I've made the change of PF_* to AF_*, thanks for explaining where the actual issue was. If you are able to try compiling the `fixes` branch again, hopefully it should work fine for you and we can then close the issue.

**Comment by ptjm** [comment]
Hi. I was able to compile the fixes branch on freebsd 11.2 using both clang and gcc

**Comment by ralight** [comment]
Thanks for checking, I'm closing the issue.
